### PR TITLE
Service Exception Front-End

### LINKED ISSUE #832: Attendance Homepage for Trainee Shows 12 AM
it should show 12 pm after 11 am.

### PR DESCRIPTION
**Pull request by iamsitting**
Trainee Select Form for Service schedules to facilitate service exceptions

also fix #832

### CODE REVIEW

File: ap/services/forms.py [hunk 1]
```diff
@@ -1,7 +1,12 @@
 from django import forms
+
 from services.models.service_hours import ServiceRoll, ServiceAttendance
 from services.models.service import Service
+import services.models.exception
```
Comment (jontey) [inline ap/services/forms.py]: should we consider renaming to ServiceException? We might run into issues in the future

File: ap/services/views.py [hunk 2]
```diff
@@ -1016,6 +1018,76 @@ def get_services_dict(self, term, week):
...
+    ctx['exceptions'] = Exception.objects.exclude(id=self.object.id)
+    ctx['form'] = self.get_modified_form(self.object)
+    ctx['button_label'] = 'Update Exception'
+    return ctx
+
+  def get_modified_form(self, obj):
+    # populate ExceptionForm with trainee ids (instead of worker ids)
+    # This is because Trainee select form uses trainee ids instead of worker ids
+    data = {}
+    data['name'] = obj.name
```
Comment (jontey) [inline ap/services/views.py]: would be nice to have the JS spread syntax in python :)

File: ap/services/views.py [hunk 3]
```diff
@@ -1016,6 +1018,76 @@ def get_services_dict(self, term, week):
...
+    return HttpResponseRedirect(self.success_url)
+
+  def get_context_data(self, **kwargs):
+    ctx = super(AddExceptionView, self).get_context_data(**kwargs)
+    ctx['exceptions'] = Exception.objects.all()
+    ctx['button_label'] = 'Add Exception'
+    return ctx
+
+
+class UpdateExceptionView(UpdateView):
```
Comment (jontey) [inline ap/services/views.py]: Maybe add GroupsRequiredMixin?

**Comment by billyeh** [comment]
Actually there was already a page/form for service exceptions in the admin, not sure if you noticed @ryanli633? Haha. I'm trying to dive into the history a bit now because I remember seeing it before. It looks identical to the one in this branch, except it also had a fancy calendar view for services which I'm not sure if we want to get working again.

**Comment by jontey** [comment]
Needs review. LGTM asap @billyeh  @yipingbenjaminliu

**Comment by yipingbenjaminliu** [comment]
@billyeh , you're up. LTGM

File: ap/aputils/groups.py [hunk 4]
```diff
@@ -31,7 +31,7 @@ def add_permissions(group, app_label_list):
 GROUP_PERMISSIONS = [
     ('training_assistant', APPS),
     ('absent_trainee_roster', ['absent_trainee_roster']),
-    ('attendance_monitors', ['attendance', 'seating', 'schedules', 'leaveslips', 'teams', 'aputils']),
+    ('attendance_monitors', ['attendance', 'seating', 'schedules', 'leaveslips', 'teams', 'aputils', 'accounts']),
```
Comment (yipingbenjaminliu) [inline ap/aputils/groups.py]: what's the reason for giving AMs access to user accounts?

File: ap/services/templates/services/services_view.html [hunk 5]
```diff
@@ -317,7 +317,12 @@ <h3>{{ category }}</h3>
   <!-- Exceptions Tab -->
   <div role="tabpanel" class="tab-pane" id="exceptions">
     <h3>Exceptions</h3>
-    <!-- TODO : Add Exceptions -->
+    <!-- TODO : Add Exceptions (In progress) -->
```
Comment (yipingbenjaminliu) [inline ap/services/templates/services/services_view.html]: is this TODO done?

File: ap/services/views.py [hunk 6]
```diff
@@ -726,6 +737,8 @@ def services_view(request, run_assign=False, generate_leaveslips=False):
   workers_bb = lJRender(WorkerIDSerializer(workers, many=True).data)
   services_bb = lJRender(ServiceCalendarSerializer(services, many=True).data)
 
+  # link to group_exceptions
```
Comment (yipingbenjaminliu) [inline ap/services/views.py]: what's this for

File: ap/aputils/groups.py (hunk 4 above)
Comment (billyeh) [reply ap/aputils/groups.py]: Reasoning is it's easier for them if they want to change someone's team 🤔 
I guess that's something we could give them on the Team admin

**Comment by billyeh** [comment]
Comments addressed, form seems to be working. I have no idea how to use it, but at least I can confirm clicking buttons puts stuff in the database and I can get it back out.

**Pull request merged**